src/Servers/Kestrel/Transport.Abstractions/src/Internal/TransportConnection.FeatureCollection.cs

```diff
@@ -100,5 +105,68 @@ void IConnectionHeartbeatFeature.OnHeartbeat(System.Action<object> action, objec
         {
             OnHeartbeat(action, state);
         }
+
+        public void OnCompleted(Func<object, Task> callback, object state)
+        {
+            if (_onCompleted == null)
+            {
+                _onCompleted = new Stack<KeyValuePair<Func<object, Task>, object>>();
+            }
+            _onCompleted.Push(new KeyValuePair<Func<object, Task>, object>(callback, state));
+        }
+
+        public Task FireOnCompleted()
+        {
+            var onCompleted = _onCompleted;
+
+            if (onCompleted == null || onCompleted.Count == 0)
+            {
+                return Task.CompletedTask;
+            }
+
+            return FireOnCompletedMayAwait(onCompleted);
+        }
+
+        private Task FireOnCompletedMayAwait(Stack<KeyValuePair<Func<object, Task>, object>> onCompleted)
+        {
+            while (onCompleted.TryPop(out var entry))
+            {
+                try
+                {
+                    var task = entry.Key.Invoke(entry.Value);
+                    if (!ReferenceEquals(task, Task.CompletedTask))
+                    {
+                        return FireOnCompletedAwaited(task, onCompleted);
+                    }
+                }
+                catch (Exception)
```

Comment:
Is there a way you'd recommend logging these exceptions? TransportConnection does not have a logger.

Reply:
Can we make TransportConnection start taking a logger?

Reply:
It could be a property.

Reply:
I think I prefer the property since that's how everything else is set on the TransportConnection. I would just make sure it doesn't nullref if the logger's not set.